lightning/src/ln/channelmanager.rs

```diff
@@ -3099,18 +3165,23 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 
 	fn claim_funds_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder<Signer>>, source: HTLCSource, payment_preimage: PaymentPreimage, forwarded_htlc_value_msat: Option<u64>, from_onchain: bool) {
 		match source {
-			HTLCSource::OutboundRoute { session_priv, .. } => {
+			HTLCSource::OutboundRoute { session_priv, mpp_id, .. } => {
 				mem::drop(channel_state_lock);
-				if {
-					let mut session_priv_bytes = [0; 32];
-					session_priv_bytes.copy_from_slice(&session_priv[..]);
-					self.pending_outbound_payments.lock().unwrap().remove(&session_priv_bytes)
-				} {
-					let mut pending_events = self.pending_events.lock().unwrap();
-					pending_events.push(events::Event::PaymentSent {
-						payment_preimage
-					});
-				} else {
+				let mut session_priv_bytes = [0; 32];
+				session_priv_bytes.copy_from_slice(&session_priv[..]);
+				let mut outbounds = self.pending_outbound_payments.lock().unwrap();
+				let found_payment = loop {
+					if let Some(mut sessions) = outbounds.remove(&mpp_id) {
+						if sessions.remove(&session_priv_bytes) {
+							self.pending_events.lock().unwrap().push(
+								events::Event::PaymentSent { payment_preimage }
```

Comment:
IIUC, an MPP will only result in one `PaymentSent` event -- let me know if I misunderstood. With that in mind, I was thinking of adding the successful path to `PaymentSent` for scoring. That way a channel's score could be improved if successfully routing through it. But if there is only one `PaymentSent` event for MPP, then such improvements will only be possible for one path. Would it be possible to include all the paths for `PaymentSent` event? Or would that require waiting for all paths to succeed before sending such an event?

Reply:
> IIUC, an MPP will only result in one PaymentSent event

Yup! We generate it as soon as one path of the MPP succeeds (i.e., we do not wait for all paths to succeed). Reasoning: as soon as we have the preimage, the rest of the paths are irrevocably "solved."

> Would it be possible to include all the paths for PaymentSent event?

Hmm, I think it's possible but not sure how pretty. Briefly looked but would have to dig more. I don't think it'd require waiting for all paths to succeed though, since we do store the path for pending payments in `HTLCSource`s

Reply:
I think as long as the receipient is "honest" (or at least wants their money and isn't buggy), we know all the paths succeeded when we get the first preimage here, so I think it would be "correct", but code-wise its probably quite tricky - the actual `HTLCSource` is stored away in the HTLC set of some other channel so not readily available here. Maybe we'll need a separate event here to indicate further paths that succeeded?

Reply:
Hmmm... yeah, that may be a good argument for separating path succeeding into its own event. Alternatively, `ChannelManager` could be parameterized by the scorer, which it would call as needed instead of using events. Though that could introduce the possibility of reentrancy. Maybe not the best idea and would be inconsistent with network updates.

Related to the [other discussion](https://github.com/rust-bitcoin/rust-lightning/pull/1053#discussion_r710621797), I guess for multi-path payments the user (or `InvoicePayer`) ultimately decides if a payment has failed when it no longer wants to retry failed paths (and hasn't received the preimage). Likewise for single-path payments. Either the user has the preimage (success) or doesn't (possible failure) and needs to decided whether to give up (i.e., no longer retry or wait).

With that in mind, maybe having a failed or successful payment at the event level isn't the right abstraction? Instead, there's `PreimageReceived`, `SuccessfulPaymentPath`, and `FailedPaymentPath`. Maybe the last two could be named in terms of HTLCs instead? Sorta thinking out loud on this, so let me know if I'm off on something.

Reply:
> Hmmm... yeah, that may be a good argument for separating path succeeding into its own event. 

I think that makes sense. It's a bit awkward to do something different for success and failure but the semantics are pretty different and Val *just* removed the separate failure event from path event. For path success the probability increases can happen later, we're not trying to make sure we get them in before retrying a new path so I think it's fine.

> Alternatively, ChannelManager could be parameterized by the scorer, which it would call as needed instead of using events. Though that could introduce the possibility of reentrancy. Maybe not the best idea and would be inconsistent with network updates.

Yea, I don't think the complexity of the reentrancy is worth it there.

> With that in mind, maybe having a failed or successful payment at the event level isn't the right abstraction? Instead, there's PreimageReceived, SuccessfulPaymentPath, and FailedPaymentPath. Maybe the last two could be named in terms of HTLCs instead? Sorta thinking out loud on this, so let me know if I'm off on something.

I think that kinda makes sense. I just suggested renaming PaymentFailed to HTLC/Path/PartFailed. For preimage receipt it does mean you need to treat the payment as sent, though, so I'm not sure as really need to swap that one.